### PR TITLE
AB#830 add microsoft azure attestation to EGo API

### PR DESCRIPTION
**Pull request by katexochen**

### CODE REVIEW

File: internal/attestation/maa_test.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+		if err := decoder.Decode(&req); err != nil {
+			http.Error(w, "could not decode json", http.StatusBadRequest)
+			return
+		}
+		report, err := base64.RawURLEncoding.DecodeString(req.Report)
+		if err != nil {
+			http.Error(w, "could not decode report", http.StatusBadRequest)
+			return
+		}
+		if string(report[:6]) != "111777" {
```
Comment (thomasten) [inline internal/attestation/maa_test.go:46]: why not check the full report?

File: internal/attestation/maa_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+			http.Error(w, "invalid report", http.StatusBadRequest)
+			return
+		}
+		if req.RuntimeData.Data == "" || req.RuntimeData.DataType == "" {
```
Comment (thomasten) [inline internal/attestation/maa_test.go:50]: check expected values here?

File: internal/attestation/maa_test.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+			http.Error(w, "missing runtime data", http.StatusBadRequest)
+			return
+		}
+		token := attestationResponse{Token: "test"}
+		response, err := json.Marshal(token)
+		if err != nil {
+			http.Error(w, "could not create reponse", http.StatusInternalServerError)
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(response)
```
Comment (thomasten) [inline internal/attestation/maa_test.go:60]:
```suggestion
		if err := json.NewEncoder(w).Encode(token); err != nil {
			http.Error(w, "could not create reponse", http.StatusInternalServerError)
		}
```

File: internal/attestation/maa_test.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	}
+	attestationProvider := httptest.NewUnstartedServer(http.HandlerFunc(createToken))
+	attestationProvider.Start()
+	defer attestationProvider.Close()
+	//
```
Comment (thomasten) [inline internal/attestation/maa_test.go:65]:
```suggestion
	}

	attestationProvider := httptest.NewUnstartedServer(http.HandlerFunc(createToken))
	attestationProvider.Start()
	defer attestationProvider.Close()

	//
```

File: internal/attestation/maa_test.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+		}
+		if string(report[:6]) != "111777" {
+			http.Error(w, "invalid report", http.StatusBadRequest)
+			return
+		}
+		if req.RuntimeData.Data == "" || req.RuntimeData.DataType == "" {
+			http.Error(w, "missing runtime data", http.StatusBadRequest)
+			return
+		}
+		token := attestationResponse{Token: "test"}
```
Comment (thomasten) [inline internal/attestation/maa_test.go:54]:
```suggestion
		}

		if string(report[:6]) != "111777" {
			http.Error(w, "invalid report", http.StatusBadRequest)
			return
		}
		if req.RuntimeData.Data == "" || req.RuntimeData.DataType == "" {
			http.Error(w, "missing runtime data", http.StatusBadRequest)
			return
		}

		token := attestationResponse{Token: "test"}
```

File: internal/attestation/maa_test.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	//
+	report := []byte("111777randombytes")
+	data := []byte("test")
+	baseurl := attestationProvider.URL
+	resp, err := CreateAzureAttestationToken(report, data, baseurl)
+	require.NoError(err)
+	require.Equal(resp, "test")
+}
+
+func TestVerifyAzureAttestationToken(t *testing.T) {
```
Comment (thomasten) [inline internal/attestation/maa_test.go:76]: add some blank lines in this test to separate logical blocks

File: internal/attestation/maa_test.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+		require.Equal(privateClaims.Debug, report.Debug)
+		require.Equal(uniqueID, report.UniqueID)
+		require.Equal(signerID, report.SignerID)
+		productID := make([]byte, 16)
+		binary.LittleEndian.PutUint16(productID, uint16(privateClaims.ProductID))
+		require.Equal(productID, report.ProductID)
+	}
+}
+
+func TestSharedProviderKeyParsing(t *testing.T) {
```
Comment (thomasten) [inline internal/attestation/maa_test.go:247]:
```suggestion
func DisabledTestSharedProviderKeyParsing(t *testing.T) {
```
Unit tests should be self-contained. Either disable this test or hardcode the response.

File: samples/azure_attestation/README.md [hunk 8]
```diff
@@ -1,13 +1,13 @@
 # Go remote attestation sample using Microsoft Azure Attestation
-This sample shows how to do remote attestation of an EGo enclave in combination with [Microsoft Azure Attestation](https://docs.microsoft.com/en-us/azure/attestation/). It consists of a server running in the enclave and a client (the relying party) that attests the server before sending a secret. The Azure Attestation Provider simplifies the attestation process for the client.
+This sample shows how to do remote attestation of an EGo enclave in combination with [Microsoft Azure Attestation](https://docs.microsoft.com/en-us/azure/attestation/). It consists of a server running in an enclave and a client (the relying party) that attests the server before sending a secret. The Azure Attestation Provider simplifies the attestation process for the client.
 
-**Note: This sample only works on SGX-FLC systems. Some error handling in this sample is omitted for brevity.**
+**Note: This sample only works on SGX-FLC systems. Some error handling is omitted for brevity.**
 
 ## How it works
 
 ![azure attestation sample](illustration.svg)
```
Comment (thomasten) [inline samples/azure_attestation/README.md:8]: why did you delete illustration.drawio?

File: samples/azure_attestation/README.md [hunk 9]
```diff
@@ -16,17 +16,17 @@ This sample shows how to do remote attestation of an EGo enclave in combination
...
+1. The client verifies the token's signature and the claims from the token body. If the token is valid and contains the correct report, the identity and integrity of the server is guranteed. The certificate is extracted from the report.
 
 The client can now establish a secure TLS connection to the enclaved server using the validated certificate and send its secret.
 
+EGo's API provides helpful functions to simplify the remote attestation with Microsoft Azure Attestation. The server can use the [CreateAzureAttestationToken()](https://pkg.go.dev/github.com/edgelesssys/ego/enclave) function form the enclave package to conduct steps 1 - 4 and get the token. The client can use the [VerifyAzureAttestationToken()](https://pkg.go.dev/github.com/edgelesssys/ego/attestation) function from EGo's attestation package to perform steps 6 and 7. While this function verifies the signature and the public claims of the token, the client has to verify the resulting report values.
```
Comment (thomasten) [inline samples/azure_attestation/README.md:28]:
```suggestion
EGo's API provides helpful functions to simplify the remote attestation with Microsoft Azure Attestation. The server can use the [CreateAzureAttestationToken()](https://pkg.go.dev/github.com/edgelesssys/ego/enclave#CreateAzureAttestationToken) function form the enclave package to conduct steps 1 - 4 and get the token. The client can use the [VerifyAzureAttestationToken()](https://pkg.go.dev/github.com/edgelesssys/ego/attestation#VerifyAzureAttestationToken) function from EGo's attestation package to perform steps 6 and 7. While this function verifies the signature and the public claims of the token, the client has to verify the resulting report values.
```

File: samples/azure_attestation/server.go [hunk 10]
```diff
@@ -1,40 +1,34 @@
...
-	report, err := enclave.GetRemoteReport(hash[:])
-	if err != nil {
-		fmt.Println(err)
-	}
-	fmt.Println("🆗 Got a report from the enclave.")
 
-	// Create Attestation Request
-	token := azureAttestation(report, cert, "https://shareduks.uks.attest.azure.net")
+	// Cerate an Azure Attestation Token.
+	token, err := enclave.CreateAzureAttestationToken(cert, attestationProviderURL)
```
Comment (thomasten) [inline samples/azure_attestation/server.go:30]: check err

File: samples/azure_attestation/ra_client/client.go [hunk 11]
```diff
@@ -3,78 +3,54 @@ package main
...
-	ProductID       uint   `json:"x-ms-sgx-product-id"`
-	SecurityVersion uint   `json:"x-ms-sgx-svn"`
-	EnclaveHeldData string `json:"x-ms-sgx-ehd"`
-}
+// attestationProviderURL is the URL of the attestation provider
+const attestationProviderURL = "https://shareduks.uks.attest.azure.net"
 
 func main() {
 	signer := flag.String("s", "", "signer ID")
-	serverAddr := flag.String("a", "localhost:8080", "server address")
```
Comment (thomasten) [inline samples/azure_attestation/ra_client/client.go:32]: let's keep this if one wants to try this remotely

**Review by thomasten — CHANGES_REQUESTED** [review]

File: samples/azure_attestation/ra_client/client.go [hunk 12]
```diff
@@ -3,78 +3,54 @@ package main
...
 
-// privateClaims are some of the private claims of an Azure Attestation token
-type privateClaims struct {
-	SignerID        string `json:"x-ms-sgx-mrsigner"`
-	ProductID       uint   `json:"x-ms-sgx-product-id"`
-	SecurityVersion uint   `json:"x-ms-sgx-svn"`
-	EnclaveHeldData string `json:"x-ms-sgx-ehd"`
-}
+// attestationProviderURL is the URL of the attestation provider
+const attestationProviderURL = "https://shareduks.uks.attest.azure.net"
```
Comment (Nirusu) [inline samples/azure_attestation/ra_client/client.go:21]: Is this URL actually regionally independent usable, or bound to the UK South region in Azure?

And if it is bound, is there any way to automatically determine the URL, or does someone running this sample on, let's say, Europe West need to change the URL manually?

File: samples/azure_attestation/README.md (hunk 8 above)
Comment (katexochen) [reply samples/azure_attestation/README.md:8]: Felix mentioned that the svg might be enough, so I tested a little and found that you can open the svg in drawio and edit it without problems, so the .drawio isn't needed.

File: samples/azure_attestation/ra_client/client.go (hunk 12 above)
Comment (katexochen) [reply samples/azure_attestation/ra_client/client.go:21]: Interesting question, I've tested this by replacing the URL with the URL of the Northern Europe Attestation Provider `https://sharedneu.neu.attest.azure.net` and the example run just fine. 

The [MAA documentation](https://docs.microsoft.com/en-us/azure/attestation/basic-concepts#regional-shared-provider) states

> Azure Attestation provides a regional shared provider in every available region. Customers can choose to use the regional shared provider for attestation, or...

To me it sounds like you should you should use the local one. But on the other hand, it shouldn't matter regarding the concept and security of MAA. So it might be okay to just use the UK South Provider for the sample.

File: samples/azure_attestation/ra_client/client.go [hunk 13]
```diff
@@ -87,60 +61,28 @@ func main() {
...
-	if err := publicClaims.Validate(jwt.Expected{Issuer: issAddr, Time: time.Now()}); err != nil {
-		return fmt.Errorf("issuer of token could not be verified: %v", err)
-	}
-	fmt.Printf("✅ Issuer of token is %s.\n", publicClaims.Issuer)
-
-	if privateClaims.SignerID != signer {
+// verifyReportValues compares the report values with that were defined in the
+// enclave.json and that were included into the binary of the server during build.
+func verifyReportValues(report attestation.Report, signer string) error {
+	if !reflect.DeepEqual(report.SignerID, []byte(signer)) {
```
Comment (thomasten) [inline samples/azure_attestation/ra_client/client.go:67]: You also need to change the sample:
```suggestion
func verifyReportValues(report attestation.Report, signer []byte) error {
	if !bytes.Equal(report.SignerID, signer) {
```
pass decoded signer (like https://github.com/edgelesssys/ego/blob/master/samples/remote_attestation/ra_client/client.go#L24)

File: samples/azure_attestation/ra_client/client.go [hunk 14]
```diff
@@ -87,60 +61,28 @@ func main() {
...
 		return errors.New("token does not contain the right signer id")
 	}
-	fmt.Println("✅ SignerID of the toke equals the SignerID you passed to the client.")
+	fmt.Println("✅ SignerID of the report equals the SignerID you passed to the client.")
 
-	if privateClaims.ProductID != 1234 {
+	productID := make([]byte, 16)
+	binary.LittleEndian.PutUint16(productID, uint16(1234))
+	if !reflect.DeepEqual(report.ProductID, productID) {
```
Comment (thomasten) [inline samples/azure_attestation/ra_client/client.go:74]:
```suggestion
	if binary.LittleEndian.Uint16(report.ProductID) != 1234 {
```

File: internal/attestation/maa_test.go [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+		t.Logf("Subtest: %v", testName)
+		//
+		// Mock attestation provider.
+		//
+		createToken := func(w http.ResponseWriter, r *http.Request) {
+			wantURL, err := url.Parse("attest/OpenEnclave?api-version=2020-10-01")
+			if err != nil {
+				http.Error(w, "could not parse wantURL", http.StatusInternalServerError)
+			}
+			if !strings.HasSuffix(r.URL.Path, wantURL.Path) || r.URL.RawQuery != wantURL.RawQuery {
```
Comment (thomasten) [inline internal/attestation/maa_test.go:60]: Why not just
```suggestion
			if r.RequestURI != "/attest/OpenEnclave?api-version=2020-10-01" {
```

File: internal/attestation/maa_test.go [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+				http.Error(w, "could not decode json", http.StatusBadRequest)
+				return
+			}
+			report, err := base64.RawURLEncoding.DecodeString(req.Report)
+			if err != nil {
+				http.Error(w, "could not decode report", http.StatusBadRequest)
+				return
+			}
+
+			if !reflect.DeepEqual(report, test.report) {
```
Comment (thomasten) [inline internal/attestation/maa_test.go:77]:
```suggestion
			if !bytes.Equal(report, test.report) {
```

**Review by thomasten — CHANGES_REQUESTED** [review]

**Review by thomasten — APPROVED** [review]
